sklearn/tree/tree.py

```diff
@@ -309,6 +309,28 @@ def fit(self, X, y, sample_weight=None, check_input=True):
 
         return self
 
+    def _validate_X_predict(self, X, check_input):
+        """Validate X whenever one tries to predict, apply, predict_proba"""
+        if self.tree_ is None:
+            raise NotFittedError("Estimator not fitted, "
+                                 "call `fit` before exploiting the model.")
+
+        if check_input:
+            X = check_array(X, dtype=DTYPE, accept_sparse="csr")
+            if issparse(X) and (X.indices.dtype != np.intc or
+                                X.indptr.dtype != np.intc):
+                raise ValueError("No support for np.int64 index based "
+                                 "sparse matrices")
+
+        n_features = X.shape[1]
```

Comment:
> But the logic would be easier if there was no check_input argument to _validate_X_predict, right?

The idea is to skip time consuming check. Note that removing this tests could lead to weird situation if `warm_start=True`.




> What do you mean by "as in the fit method"? The idea of check_input is that the check is run exactly once in the ensembles.

For instance, parameter values are checked before fitting any tree in the forest. This is explicit in the decision tree fit method.